ouroboros-consensus/src/Ouroboros/Consensus/Storage/ChainDB/Impl/LgrDB.hs

```diff
@@ -357,13 +402,15 @@ validate :: forall m blk. (IOLike m, LedgerSupportsProtocol blk, HasCallStack)
          -> m (ValidateResult blk)
 validate LgrDB{..} ledgerDB blockCache numRollbacks trace = \hdrs -> do
     aps <- mkAps hdrs <$> atomically (readTVar varPrevApplied)
-    res <- fmap rewrap $ LedgerDB.defaultResolveWithErrors resolveBlock $
-             LedgerDB.ledgerDbSwitch
-               (configLedgerDb cfg)
-               numRollbacks
-               (lift . lift . trace)
-               aps
-               ledgerDB
+    res <- fmap rewrap $
+             LedgerDB.defaultReadKeySets (readKeySets lgrOnDiskLedgerStDb) $
+             LedgerDB.defaultResolveWithErrors (LedgerDB.DbReader . lift . resolveBlock) $
+               LedgerDB.ledgerDbSwitch
+                 (configLedgerDb cfg)
+                 numRollbacks
+                 (lift . lift . LedgerDB.DbReader . lift . trace)
```

Comment:
I'm not happy with this monad stack madness. I wonder if we should refactor this a bit.